From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lee Chagolla-Christensen <lee@localhost>
Date: Sat, 13 Dec 2025 10:34:29 -0800
Subject: [PATCH] 8773: Add option to toggle Tool toolbar

add option to toggle Tool toolbar

Ensure frontend sets default options

Allows the frontend tool to call setDefaultOptionValues.
Without this, on startup, the frontend tool (the main project
window) will not use toolbar configured toolbar setting.
---
 .../help/topics/Tool/ToolOptions_Dialog.htm   |  6 ++++++
 .../java/docking/DockingWindowManager.java    | 20 +++++++++++++++++++
 .../java/docking/WindowActionManager.java     |  9 ++++++++-
 .../ghidra/framework/main/FrontEndTool.java   |  1 +
 .../framework/plugintool/PluginTool.java      | 10 +++++++++-
 5 files changed, 44 insertions(+), 2 deletions(-)

diff --git a/Ghidra/Features/Base/src/main/help/help/topics/Tool/ToolOptions_Dialog.htm b/Ghidra/Features/Base/src/main/help/help/topics/Tool/ToolOptions_Dialog.htm
index 574d2d8fc1..e3e1e358f5 100644
--- a/Ghidra/Features/Base/src/main/help/help/topics/Tool/ToolOptions_Dialog.htm
+++ b/Ghidra/Features/Base/src/main/help/help/topics/Tool/ToolOptions_Dialog.htm
@@ -330,6 +330,12 @@
                 <TD valign="top" align="left">If selected, a program tab will be displayed even
                 there is only one program open.</TD>
               </TR>
+              <TR valign="middle">
+                <TD valign="top" width="200" align="left">Show Toolbar</TD>
+
+                <TD valign="top" align="left">If selected, the main toolbar will be visible at
+                the top of the tool window.</TD>
+              </TR>
               <TR valign="middle">
                 <TD valign="top" width="200" align="left">Subroutine Model</TD>
 
diff --git a/Ghidra/Framework/Docking/src/main/java/docking/DockingWindowManager.java b/Ghidra/Framework/Docking/src/main/java/docking/DockingWindowManager.java
index 163a6b6130..71a6c56639 100644
--- a/Ghidra/Framework/Docking/src/main/java/docking/DockingWindowManager.java
+++ b/Ghidra/Framework/Docking/src/main/java/docking/DockingWindowManager.java
@@ -106,6 +106,7 @@ public class DockingWindowManager implements PropertyChangeListener, Placeholder
 	private boolean hasStatusBar;
 
 	private boolean windowsOnTop;
+	private boolean toolbarVisible = true;
 
 	private Window lastActiveWindow;
 
@@ -2294,6 +2295,25 @@ public class DockingWindowManager implements PropertyChangeListener, Placeholder
 		return windowsOnTop;
 	}
 
+	/**
+	 * Sets the visibility of the main toolbar.
+	 *
+	 * @param visible true to show the toolbar, false to hide it.
+	 */
+	public void setToolbarVisible(boolean visible) {
+		this.toolbarVisible = visible;
+		scheduleUpdate();
+	}
+
+	/**
+	 * Returns true if the toolbar is visible.
+	 *
+	 * @return true if the toolbar is visible.
+	 */
+	public boolean isToolbarVisible() {
+		return toolbarVisible;
+	}
+
 	/**
 	 * Returns a list with all the windows in the windowStack. Used for testing.
 	 *
diff --git a/Ghidra/Framework/Docking/src/main/java/docking/WindowActionManager.java b/Ghidra/Framework/Docking/src/main/java/docking/WindowActionManager.java
index 7c30d33c91..6b78de618b 100644
--- a/Ghidra/Framework/Docking/src/main/java/docking/WindowActionManager.java
+++ b/Ghidra/Framework/Docking/src/main/java/docking/WindowActionManager.java
@@ -31,6 +31,7 @@ public class WindowActionManager {
 	private MenuBarManager menuBarMgr;
 	private ToolBarManager toolBarMgr;
 	private final WindowNode node;
+	private final DockingWindowManager winMgr;
 
 	private boolean disposed;
 
@@ -56,6 +57,7 @@ public class WindowActionManager {
 			DockingWindowManager winMgr, MenuGroupMap menuGroupMap) {
 
 		this.node = node;
+		this.winMgr = winMgr;
 		actionToProxyMap = new HashMap<>();
 		menuBarMgr = new MenuBarManager(menuBarHandler, menuGroupMap);
 		toolBarMgr = new ToolBarManager(winMgr);
@@ -100,7 +102,12 @@ public class WindowActionManager {
 			node.setMenuBar(menuBar);
 		}
 
-		node.setToolBar(toolBarMgr.getToolBar());
+		if (winMgr.isToolbarVisible()) {
+			node.setToolBar(toolBarMgr.getToolBar());
+		}
+		else {
+			node.setToolBar(null);
+		}
 		node.validate();
 	}
 
diff --git a/Ghidra/Framework/Project/src/main/java/ghidra/framework/main/FrontEndTool.java b/Ghidra/Framework/Project/src/main/java/ghidra/framework/main/FrontEndTool.java
index 71a2ee92f6..97ce05ef98 100644
--- a/Ghidra/Framework/Project/src/main/java/ghidra/framework/main/FrontEndTool.java
+++ b/Ghidra/Framework/Project/src/main/java/ghidra/framework/main/FrontEndTool.java
@@ -744,6 +744,7 @@ public class FrontEndTool extends PluginTool implements OptionsChangeListener {
 		Element elem = root.getChild("TOOL");
 
 		restoreOptionsFromXml(elem);
+		setDefaultOptionValues(); // Apply loaded option values to window manager
 		try {
 			restorePluginsFromXml(elem);
 		}
diff --git a/Ghidra/Framework/Project/src/main/java/ghidra/framework/plugintool/PluginTool.java b/Ghidra/Framework/Project/src/main/java/ghidra/framework/plugintool/PluginTool.java
index 178c825bb0..5aea859b84 100644
--- a/Ghidra/Framework/Project/src/main/java/ghidra/framework/plugintool/PluginTool.java
+++ b/Ghidra/Framework/Project/src/main/java/ghidra/framework/plugintool/PluginTool.java
@@ -96,6 +96,7 @@ public abstract class PluginTool extends AbstractDockingTool {
 	public final static String PLUGIN_COUNT_PROPERTY_NAME = "PluginCount";
 
 	private static final String DOCKING_WINDOWS_ON_TOP = "Docking Windows On Top";
+	private static final String SHOW_TOOLBAR = "Show Toolbar";
 
 	private static final String SAVE_DIALOG_TITLE = "Save Tool - Possible Conflict";
 
@@ -266,10 +267,12 @@ public abstract class PluginTool extends AbstractDockingTool {
 		// placeholder
 	}
 
-	private void setDefaultOptionValues() {
+	protected void setDefaultOptionValues() {
 		Options toolOptions = optionsMgr.getOptions(ToolConstants.TOOL_OPTIONS);
 		boolean windowsOnTop = toolOptions.getBoolean(DOCKING_WINDOWS_ON_TOP, false);
 		winMgr.setWindowsOnTop(windowsOnTop);
+		boolean showToolbar = toolOptions.getBoolean(SHOW_TOOLBAR, true);
+		winMgr.setToolbarVisible(showToolbar);
 	}
 
 	private void initOptions() {
@@ -277,6 +280,8 @@ public abstract class PluginTool extends AbstractDockingTool {
 		toolOptions.registerOption(DOCKING_WINDOWS_ON_TOP, false, null,
 			"Determines whether a docked window will always be shown on " +
 				"top of its parent window.");
+		toolOptions.registerOption(SHOW_TOOLBAR, true, null,
+			"Determines whether the main toolbar is visible.");
 
 		// we must call this before the init work below to make sure that the options object
 		// we use has been created
@@ -291,6 +296,9 @@ public abstract class PluginTool extends AbstractDockingTool {
 		if (name.equals(DOCKING_WINDOWS_ON_TOP)) {
 			winMgr.setWindowsOnTop(((Boolean) newValue).booleanValue());
 		}
+		else if (name.equals(SHOW_TOOLBAR)) {
+			winMgr.setToolbarVisible(((Boolean) newValue).booleanValue());
+		}
 	}
 
 	/**
-- 
2.45.1

